pkg/scheduler/queue/tree_queue.go

```diff
@@ -0,0 +1,282 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package queue
+
+import (
+	"container/list"
+)
+
+type QueuePath []string //nolint:revive
+type QueueIndex int     //nolint:revive
+
+const localQueueIndex = -1
+
+// TreeQueue is a hierarchical queue implementation with an arbitrary amount of child queues.
+//
+// TreeQueue internally maintains round-robin fair queuing across all of its queue dimensions.
+// Each queuing dimension is modeled as a node in the tree, internally reachable through a QueuePath.
+//
+// The QueuePath is an ordered array of strings describing the path through the tree to the node,
+// which contains the FIFO local queue of all items enqueued for that queuing dimension.
+//
+// When dequeuing from a given node, the node will round-robin equally between dequeuing directly
+// from its own local queue and dequeuing recursively from its list of child TreeQueues.
```

Comment:
Would it be valid for a node to both have child nodes and also have items in its queue? Seems like only a leaf node should have items in its queue, and if this was the case, it would help simplify some of the dequeuing logic.

Reply:
> Would it be valid for a node to both have child nodes and also have items in its queue?

Conceptually, yes though I don't think we plan to use it in Mimir in that way.

If a queue path is a sort of logical category, a node's local queue contains items that cannot be further subcategorized. This is a recursive truth the holds all the way down to the leaf, and the dequeuing consumer of this structure never has to know the internals.

For Mimir, the consumer asks for a request for `tenantA`, it gets a request for `tenantA`. As it continues to ask for requests for `tenantA`, it is always receiving a request that has been fairly dequeued in round robin fashion for the 0 to N possible subcategories of "tenant-A".
The consumer is always getting something from the category it wants _or its subcategories_, and it doesn't even have to know if any subcategories exist at all.

To answer
> if we only dequeue from leaf nodes could this be simpler (paraphrasing)

I feel pretty confident this is not the case.
We currently get to treat every node as if it is exactly the same, which is a luxury.

We currently do a single integer comparison at each node traversed during a dequeue to decide whether to dequeue from local queue or recur into a child queue and perform identical operations.
While I won't leave a dissertation of all the alternatives here, I have considered this extensively and I do not see a structure that is any less awkward than this.